File: ios/MitiMaiti/ViewModels/ProfileViewModel.swift
import SwiftUI

@MainActor
class ProfileViewModel: ObservableObject {
    @Published var user: User = {
        var u = MockData.currentUser
        // If the user set their name during onboarding, use it as the
        // profile name instead of the mock default.
        let storedName = UserProfileStore.shared.firstName
        if !storedName.isEmpty { u.displayName = storedName }
        return u
    }()
    @Published var isLoading = false
    @Published var error: String?
    @Published var isSaving = false
    @Published var saveSuccess = false

    // Edit fields
    @Published var editBio = ""
    @Published var editHeight = ""
    @Published var editEducation = ""
    @Published var editOccupation = ""
    @Published var editCompany = ""
    @Published var editReligion = ""
    @Published var editSmoking = ""
    @Published var editDrinking = ""
    @Published var editExercise = ""
    @Published var editFluency: SindhiFluency = .fluent
    @Published var editFamilyValues: FamilyValues = .moderate
    @Published var editFoodPreference: FoodPreference = .vegetarian

    private let api = APIService.shared

    var profileStats: [(String, String, String)] {
        [
            ("eye.fill", "47", "Views"),
            ("heart.fill", "12", "Likes"),
            ("person.2.fill", "5", "Matches")
        ]
    }

    /// Computed profile completeness — reflects what's actually filled on the
    /// ProfileViewModel + the user's stored photos/prompts, rather than a
    /// static value on the User model. Recomputes whenever any @Published
    /// edit field changes (ObservableObject triggers view refresh).
    var computedCompleteness: Int {
        var filled = 0
        let total = 15 // photos + bio + 8 basics + prompts + 4 sindhi/identity

        // Photos
        if !user.photos.isEmpty { filled += 1 }
        // Prompts
        if !user.prompts.isEmpty { filled += 1 }
        // Bio
        if !editBio.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        // Basics (8)
        if !editHeight.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        if !editEducation.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        if !editOccupation.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        if !editCompany.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        if !editReligion.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        if !editSmoking.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        if !editDrinking.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        if !editExercise.trimmingCharacters(in: .whitespaces).isEmpty { filled += 1 }
        // Sindhi / Family identity (4)
        if user.sindhiFluency != nil { filled += 1 }
        if user.familyValues != nil { filled += 1 }
        if user.foodPreference != nil { filled += 1 }
        if !(user.gotra ?? "").isEmpty { filled += 1 }

        return Int((Double(filled) / Double(total)) * 100.0)
    }

    func loadProfile() {
        isLoading = true
        Task {
            do {
                var fetched = try await api.fetchProfile()
                let storedName = UserProfileStore.shared.firstName
                if !storedName.isEmpty { fetched.displayName = storedName }
                user = fetched
                populateEditFields()
                isLoading = false
            } catch {
                self.error = error.localizedDescription
                isLoading = false
            }
        }
    }

    func populateEditFields() {
        editBio = user.bio ?? ""
        editHeight = user.heightCm.map(String.init) ?? ""
        editEducation = user.education ?? ""
        editOccupation = user.occupation ?? ""
        editCompany = user.company ?? ""
        editReligion = user.religion ?? ""
        editSmoking = user.smoking ?? ""
        editDrinking = user.drinking ?? ""
        editExercise = user.exercise ?? ""
        editFluency = user.sindhiFluency ?? .fluent
        editFamilyValues = user.familyValues ?? .moderate
        editFoodPreference = user.foodPreference ?? .vegetarian
    }

    func saveProfile() {
        isSaving = true
        Task {
            do {
                _ = try await api.updateProfile([:])
                isSaving = false
                saveSuccess = true

                try? await Task.sleep(nanoseconds: 2_000_000_000)
                saveSuccess = false
            } catch {
                self.error = error.localizedDescription
                isSaving = false
            }
        }
    }
}

File: android/app/src/main/java/com/mitimaiti/app/viewmodels/ProfileViewModel.kt
package com.mitimaiti.app.viewmodels

import android.net.Uri
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mitimaiti.app.models.*
import com.mitimaiti.app.services.APIService
import com.mitimaiti.app.services.PhotoRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class ProfileStats(val views: Int = 0, val likes: Int = 0, val matches: Int = 0)

class ProfileViewModel : ViewModel() {
    private val _user = MutableStateFlow<User?>(null)
    val user: StateFlow<User?> = _user.asStateFlow()
    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()
    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()
    private val _isSaving = MutableStateFlow(false)
    val isSaving: StateFlow<Boolean> = _isSaving.asStateFlow()
    private val _saveSuccess = MutableStateFlow(false)
    val saveSuccess: StateFlow<Boolean> = _saveSuccess.asStateFlow()

    // Basic fields
    val editBio = MutableStateFlow("")
    val editHeight = MutableStateFlow<Int?>(null)
    val editEducation = MutableStateFlow("")
    val editOccupation = MutableStateFlow("")
    val editCompany = MutableStateFlow("")
    val editReligion = MutableStateFlow("")

    // Lifestyle fields
    val editSmoking = MutableStateFlow("")
    val editDrinking = MutableStateFlow("")
    val editExercise = MutableStateFlow("")
    val editWantKids = MutableStateFlow("")
    val editSettlingTimeline = MutableStateFlow("")

    // Sindhi Identity fields
    val editFluency = MutableStateFlow<SindhiFluency?>(null)
    val editDialect = MutableStateFlow("")
    val editGeneration = MutableStateFlow("")
    val editGotra = MutableStateFlow("")
    val editFamilyOriginCity = MutableStateFlow("")
    val editFamilyOriginCountry = MutableStateFlow("")
    val editCommunitySubGroup = MutableStateFlow("")
    val editMotherTongue = MutableStateFlow("")

    // Cultural fields
    val editFamilyValues = MutableStateFlow<FamilyValues?>(null)
    val editFoodPreference = MutableStateFlow<FoodPreference?>(null)
    val editFestivals = MutableStateFlow<List<String>>(emptyList())
    val editCuisinePreferences = MutableStateFlow<List<String>>(emptyList())

    // Personality fields
    val editInterests = MutableStateFlow<List<String>>(emptyList())
    val editMusicPreferences = MutableStateFlow<List<String>>(emptyList())
    val editMovieGenres = MutableStateFlow<List<String>>(emptyList())
    val editTravelStyle = MutableStateFlow("")
    val editLanguages = MutableStateFlow<List<String>>(emptyList())

    // Prompts editing
    val editPrompts = MutableStateFlow<List<UserPrompt>>(emptyList())

    // Photos from PhotoRepository (shared with onboarding)
    val userPhotos: StateFlow<List<Uri>> = PhotoRepository.photos
    val primaryPhotoUri: Uri? get() = PhotoRepository.primaryPhotoUri

    fun addPhoto(uri: Uri) { PhotoRepository.addPhoto(uri) }
    fun removePhoto(index: Int) { PhotoRepository.removePhoto(index) }
    fun setPrimaryPhoto(index: Int) { PhotoRepository.setPrimaryPhoto(index) }

    val profileStats = ProfileStats(views = 142, likes = 38, matches = 12)

    fun loadProfile() {
        viewModelScope.launch {
            _isLoading.value = true
            APIService.fetchProfile()
                .onSuccess { _user.value = it; populateEditFields(it) }
                .onFailure { _error.value = "Failed to load profile" }
            _isLoading.value = false
        }
    }

    private fun populateEditFields(user: User) {
        editBio.value = user.bio
        editHeight.value = user.heightCm
        editEducation.value = user.education ?: ""
        editOccupation.value = user.occupation ?: ""
        editCompany.value = user.company ?: ""
        editReligion.value = user.religion ?: ""
        editSmoking.value = user.smoking ?: ""
        editDrinking.value = user.drinking ?: ""
        editExercise.value = user.exercise ?: ""
        editWantKids.value = user.wantKids ?: ""
        editSettlingTimeline.value = user.settlingTimeline ?: ""
        editFluency.value = user.sindhiFluency
        editDialect.value = user.sindhiDialect ?: ""
        editGeneration.value = user.generation ?: ""
        editGotra.value = user.gotra ?: ""
        editFamilyOriginCity.value = user.familyOriginCity ?: ""
        editFamilyOriginCountry.value = user.familyOriginCountry ?: ""
        editCommunitySubGroup.value = user.communitySubGroup ?: ""
        editMotherTongue.value = user.motherTongue ?: ""
        editFamilyValues.value = user.familyValues
        editFoodPreference.value = user.foodPreference
        editFestivals.value = user.festivalsCelebrated
        editCuisinePreferences.value = user.cuisinePreferences
        editInterests.value = user.interests
        editMusicPreferences.value = user.musicPreferences
        editMovieGenres.value = user.movieGenres
        editTravelStyle.value = user.travelStyle ?: ""
        editLanguages.value = user.languages
        editPrompts.value = user.prompts
    }

    fun saveProfile() {
        viewModelScope.launch {
            _isSaving.value = true
            APIService.updateProfile(mapOf("bio" to editBio.value))
                .onSuccess {
                    _saveSuccess.value = true
                    _user.value = _user.value?.copy(
                        bio = editBio.value,
                        heightCm = editHeight.value,
                        education = editEducation.value.ifBlank { null },
                        occupation = editOccupation.value.ifBlank { null },
                        company = editCompany.value.ifBlank { null },
                        religion = editReligion.value.ifBlank { null },
                        smoking = editSmoking.value.ifBlank { null },
                        drinking = editDrinking.value.ifBlank { null },
                        exercise = editExercise.value.ifBlank { null },
                        wantKids = editWantKids.value.ifBlank { null },
                        settlingTimeline = editSettlingTimeline.value.ifBlank { null },
                        sindhiFluency = editFluency.value,
                        sindhiDialect = editDialect.value.ifBlank { null },
                        generation = editGeneration.value.ifBlank { null },
                        gotra = editGotra.value.ifBlank { null },
                        familyOriginCity = editFamilyOriginCity.value.ifBlank { null },
                        familyOriginCountry = editFamilyOriginCountry.value.ifBlank { null },
                        communitySubGroup = editCommunitySubGroup.value.ifBlank { null },
                        motherTongue = editMotherTongue.value.ifBlank { null },
                        familyValues = editFamilyValues.value,
                        foodPreference = editFoodPreference.value,
                        festivalsCelebrated = editFestivals.value,
                        cuisinePreferences = editCuisinePreferences.value,
                        interests = editInterests.value,
                        musicPreferences = editMusicPreferences.value,
                        movieGenres = editMovieGenres.value,
                        travelStyle = editTravelStyle.value.ifBlank { null },
                        languages = editLanguages.value,
                        prompts = editPrompts.value
                    )
                }
                .onFailure { _error.value = "Failed to save profile" }
            _isSaving.value = false
        }
    }

    fun dismissSaveSuccess() { _saveSuccess.value = false }
}
